File: MediaSundaySchedule/mediaSchedule.py
from datetime import datetime, timedelta
from constraint import Problem, AllDifferentConstraint, InSetConstraint

def generate_schedule(team_data, start_date, end_date):
    # Create a list of dates from start_date to end_date
    dates = [start_date + timedelta(7 * i) for i in range((end_date - start_date).days // 7 + 1)]

    talents_needed = {
        "photography": 4,
        "video": 2,
        "visionmixing": 1,
        "livestream": 1
    }

    talents_combinations = [["photography", "video", "visionmixing", "livestream"]] * len(dates)

    # Create a problem instance
    problem = Problem()

    # Add variables for each date with the possible talents combinations
    for i, date in enumerate(dates):
        problem.addVariable(f"date_{i}", talents_combinations[i])

    # Add constraint: each date's talents must be unique
    problem.addConstraint(AllDifferentConstraint(), [f"date_{i}" for i in range(len(dates))])

    # Add constraint: the count of each talent must be as required
    for talent, required_count in talents_needed.items():
        problem.addConstraint(lambda *args: sum(1 for arg in args if arg == talent) == required_count, [f"date_{i}" for i in range(len(dates))])

    # Add constraint: members cannot be scheduled for two talents on the same date
    for member in team_data.keys():
        talents_available = team_data[member]["talents"]
        for i in range(len(dates)):
            problem.addConstraint(InSetConstraint([talents_available]), [f"date_{i}"])

    # Add constraint: members cannot be scheduled for consecutive Sundays
    for member in team_data.keys():
        for i in range(len(dates) - 1):
            problem.addConstraint(lambda d1, d2, member=member: abs((d2 - d1).days) != 7 or d2 == None, (f"date_{i}", f"date_{i+1}", member))

    # Solve the problem and get the solution
    solution = problem.getSolution()

    if solution is None:
        print("No valid schedule found.")
        return None

    # Convert solution keys to datetime objects
    schedule = {datetime.strptime(key.split('_')[1], "%Y-%m-%d"): value for key, value in solution.items()}

    return schedule

# Sample team data (you can replace this with the actual data)
team_data = {
    "Member 1": {"talents": ["visionmixing", "video"], "preferences": ["1st", "2nd"]},
    "Member 2": {"talents": ["photography", "livestream"], "preferences": ["1st"]},
    "Member 3": {"talents": ["video"], "preferences": ["2nd"]},
    # Add more team members here
}

start_date = datetime(2023, 8, 6)  # Start date for August 2023 (1st Sunday)
end_date = datetime(2023, 8, 27)  # End date for August 2023 (4th Sunday)

schedule = generate_schedule(team_data, start_date, end_date)
print(schedule)



# # Sample team data (you can replace this with the actual data)
# team_data = {
#     "Member 1": {"talents": ["visionmixing", "video"], "preferences": ["1st", "2nd"]},
#     "Member 2": {"talents": ["photography", "livestream"], "preferences": ["1st"]},
#     "Member 3": {"talents": ["video"], "preferences": ["2nd"]},
#     "Member 4": {"talents": ["visionmixing", "video"], "preferences": ["1st", "2nd"]},
#     "Member 5": {"talents": ["photography", "livestream"], "preferences": ["2nd"]},
#     "Member 6": {"talents": ["visionmixing", "photography"], "preferences": ["1st", "2nd"]},
#     "Member 7": {"talents": ["photography"], "preferences": ["1st", "2nd"]},
#     "Member 8": {"talents": ["visionmixing", "livestream"], "preferences": ["1st", "2nd"]},
#     # Add more team members here
# }

File: MediaSundaySchedule/mediaSchedule.js
function createSchedule(teamData, startDate, endDate) {
  // Define the required roles for each service
  const serviceRequirements = {
    '1st': { photography: 2, video: 2, livestream: 1, visionmixing: 1 },
    '2nd': { photography: 2, video: 2 }
  };

  let currentDate = new Date(startDate);
  let schedule = {};
  
  // Function to add days to a date
  function addDays(date, days) {
    let result = new Date(date);
    result.setDate(result.getDate() + days);
    return result;
  }

  // Track the last time a member was scheduled to ensure rest
  let lastScheduled = {};

  while (currentDate <= endDate) {
    let dateKey = currentDate.toISOString().split('T')[0]; // YYYY-MM-DD format
    schedule[dateKey] = { '1st': [], '2nd': [] };

    Object.entries(serviceRequirements).forEach(([service, roles]) => {
      Object.entries(roles).forEach(([role, numNeeded]) => {
        let assigned = 0;

        // Sort team members by the last scheduled date to prioritize those who rested more
        let candidates = teamData.sort((a, b) => {
          let lastA = lastScheduled[a.name] || new Date(startDate);
          let lastB = lastScheduled[b.name] || new Date(startDate);
          return lastA - lastB;
        });

        for (let member of candidates) {
          if (assigned < numNeeded) {
            // Check if member can serve in this role and service, considering consecutive scheduling
            let canServe = (!lastScheduled[member.name] || addDays(lastScheduled[member.name], 7) <= currentDate || 
                            (Object.keys(lastScheduled).length / teamData.length) >= 2) && // Allow for consecutive scheduling if necessary
                            (member.talents.includes(role) && member.preferences.includes(service));

            if (canServe) {
              schedule[dateKey][service].push({ name: member.name, talent: role });
              lastScheduled[member.name] = currentDate; // Update last scheduled
              assigned++;
            }
          } else {
            break; // Break early if all needed roles are filled
          }
        }
      });
    });

    currentDate = addDays(currentDate, 7); // Move to the next Sunday
  }

  return schedule;
}


const teamData = [
  // {name:'Jose', talents: ['visionmixing', 'photography'], preferences: ['1st', '2nd'] },
  {name:'Chudu', talents: ['photography'], preferences: ['1st'] },
  // {name:'Chisha', talents: ['photography'], preferences: ['1st'] },
  {name:'Charity', talents: ['visionmixing'], preferences: ['1st', '2nd'] },
  {name:'Felix', talents: ['photography','video'], preferences: ['1st','2nd'] },
  // {name:'Nobu', talents: ['photography', 'livestream'], preferences: ['1st', '2nd'] },
  {name:'Thokozile', talents: ['photography', 'video', 'livestream'], preferences: ['1st','2nd'] },
  {name:'Prudence', talents: ['photography, livestream'], preferences: ['1st'] },
  // {name:'Clement', talents: ['visionmixing', 'video'], preferences: ['1st', '2nd'] },
  // {name:'Mara', talents: ['livestream'], preferences: ['1st','2nd'] },
  {name:'Kondwani', talents: ['livestream'], preferences: ['1st'] },
  {name:'Eneya', talents: ['visionmixing', 'video'], preferences: ['1st'] },
  {name:'Yande', talents: ['video','photography', 'visionmixing'], preferences: ['1st','2nd'] },
  {name:'Chanda', talents: ['photography','video'], preferences: ['2nd'] },
  {name:'Lombe', talents: ['visionmixing', 'video'], preferences: ['1st'] },
  {name:'Emmanuel Luks', talents: ['visionmixing','photography'], preferences: ['1st', '2nd'] },
  // {name:'Medelina', talents: ['photography', 'video'], preferences: ['1st', '2nd'] },
  {name:'Sarah', talents: ['visionmixing', 'video','photography'], preferences: ['1st'] },
  // {name:'Barnabas', talents: ['video'], preferences: ['2nd'] },
  // {name:'Lukundo', talents: ['photography'], preferences: ['2nd'] },
  // {name:'Bright', talents: ['photography', 'livestream'], preferences: ['1st', '2nd'] },
  {name:'Tabo', talents: ['photography','livestream'], preferences: ['1st'] },
  // {name:'Niza', talents: ['video', 'livestream'], preferences: ['1st', '2nd'] },
  {name:'Mukutwa', talents: ['photography'], preferences: ['2nd'] },
  // {name:'Mike', talents: ['photography'], preferences: ['2nd'] },
  {name:'David Michael', talents: ['video', 'photography'], preferences: ['1st', '2nd'] },
  // {name:'Chibwe', talents: ['photography'], preferences: ['1st'] },
  {name:'Kusibili', talents: ['photography', 'video'], preferences: ['1st','2nd'] }
  // 'Yombo': { talents: ['photography'], preferences: ['1st', '2nd'] },
   // 'Ossie': { talents: ['video', 'livestream'], preferences: ['1st'] }, //is travelling
];

const startDate = '2024-03-01';
const endDate = '2024-03-31';

const schedule = createSchedule(teamData, new Date(startDate), new Date(endDate));
// console.log(JSON.stringify(schedule, null, 2));

Object.entries(schedule).forEach(([date, services]) => {
  console.log(`Date: ${date}`);
  Object.entries(services).forEach(([service, assignments]) => {
    console.log(`  Service: ${service}`);
    assignments.forEach((assignment, index) => {
      console.log(`    ${index + 1}. Name: ${assignment.name}, Talent: ${assignment.talent}`);
    });
  });
});